micro_http/src/connection.rs

```diff
@@ -0,0 +1,714 @@
+                    } else {
+                        response_fully_written = true;
+                    }
+                }
+                Err(_) => {
+                    return Err(ConnectionError::StreamError);
+                }
+            }
+        }
+
+        if response_fully_written {
+            self.response_buffer.take();
+        }
+
+        Ok(())
+    }
+
+    /// Send a response back to the source of a request.
+    pub fn send_response(&mut self, response: Response) {
+        self.response_queue.push_back(response);
+    }
+
+    fn shift_buffer_left(&mut self, line_start_index: usize, end_cursor: usize) {
+        // We don't want to shift something that is already at the beginning.
+        if line_start_index != 0 {
+            // Move the bytes from `line_start_index` to the beginning of the buffer.
+            for cursor in 0..(end_cursor - line_start_index) {
+                self.buffer[cursor] = self.buffer[line_start_index + cursor];
+            }
+
+            // Clear the rest of the buffer.
+            for cursor in (end_cursor - line_start_index)..end_cursor {
+                self.buffer[cursor] = 0;
+            }
+        }
+
+        // Update `read_cursor`.
+        self.read_cursor = end_cursor - line_start_index;
+    }
+
+    /// Returns the first parsed request in the queue.
+    pub fn pop_parsed_request(&mut self) -> Option<Request> {
+        self.parsed_requests.pop_front()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::Version;
+    use super::*;
+    use std::os::unix::net::UnixStream;
+
+    #[test]
+    #[allow(clippy::cognitive_complexity)]
+    fn test_try_read_connection() {
+        // Test request with `Expect` header.
+        if let Ok((mut sender, receiver)) = UnixStream::pair() {
+            receiver.set_nonblocking(true).expect("Can't modify socket");
+            let mut conn = HttpConnection::new(receiver);
+            sender
+                .write_all(
+                    b"PATCH http://localhost/home HTTP/1.1\r\n \
+                                     Expect: 100-continue\r\n \
+                                     Content-Length: 26\r\n \
+                                     Transfer-Encoding: chunked\r\n\r\n",
+                )
+                .unwrap();
+            assert!(conn.try_read().is_ok());
+
+            sender.write_all(b"this is not\n\r\na json \nbody").unwrap();
+            conn.try_read().unwrap();
+            let request = conn.pop_parsed_request().unwrap();
+            let expected_request = Request::try_from(
+                b"PATCH http://localhost/home HTTP/1.1\r\n \
+                                     Expect: 100-continue\r\n \
+                                     Transfer-Encoding: chunked\r\n \
+                                     Content-Length: 26\r\n\r\nthis is not\n\r\na json \nbody",
+            )
+            .unwrap();
+            assert_eq!(request, expected_request);
```

Comment:
Testing that the code from `try_read` works by comparing the result with what `try_from` is not ideal. You're testing one functionality using another functionality which is also written in this module and which might actually go away in a future refatoring if we switch MMDS to use try_read.This is not a reliable way of checking the expected request.

The expected output should be static. Because of this I suggest you create a request object in this test function. This should be easy to achieve as I notice that all Request fields are public anyway.